private/dynamicHillClimb.m

```diff
@@ -0,0 +1,391 @@
+%   ExpandFactor      : factor of expansion upon success (default 2)
+%   ContractFactor    : factor of contraction upon failure (default 0.45)
+%   StuckSearchFactor : how far to expand again after got stuck (default 4)
+%   Barrier           : use barrier on bounds (default extreme barrier)
+%   Display           : off|iter|debug, text output (default off)
+%
+% Output:
+% x   : best guess for parameters
+% fval: objective function at the solution, generally fval=fun(x)
+% exitflag:
+%   1 : The function converged to a solution x
+%   0 : Number of function evaluations exceeded options.MaxFunEvals.
+%   -1: The algorithm was terminated inappropriately
+% output : struct with meta information:
+%   iterations  : number of iterations
+%   funcCount   : number of function evaluations
+%   algorithm   : name of the algorithm
+%   t_cpu       : cpu time
+%
+% History:
+% 2017/09/27 Yannik Schaelte
+
+% number of variables
+dim  = length(x0); % number of variables
+
+% interpret options
+[tolX,tolFun,maxFunEvals,outputFcn,...
+    initialStepSize,expandFactor,contractFactor,stuckSearchFactor,barrier,...
+    display]...
+    = f_extractOptions(options,dim);
+if (isa(outputFcn,'function_handle'))
+    visual = true;
+else
+    visual = false;
+end
+
+% create column vectors
+lb      = lb(:);
+ub      = ub(:);
+x0      = x0(:);
+normalize   = @(x) f_normalize(x,lb,ub);
+denormalize = @(y) f_denormalize(y,lb,ub);
+y0      = normalize(x0);
+tolY    = tolX / norm(ub-lb);
+% max step size
+vmax = initialStepSize * ones(dim,1);
+
+% wrap function to consider boundaries
+fun = @(y,funEvals) f_wrap_fun(denormalize(y),fun,lb,ub,barrier,funEvals,maxFunEvals);
+
+% init run variables
+smax   = [vmax;-1];        % array of max step sizes, extra value for extra vector
+[step,norms]   = f_init_step(vmax); % matrix of step vectors
+xstep  = step;             % steps before last motion
+xnorms = norms;
+gradv  = zeros(dim,1);     % gradient vector
+gradi  = -1;               % index of gradient vector, -1 indicates gradv is not set
+lastj  = -1;               % index of last step taken
+stuck  = false;            % is process stuck (in min/max)? set when step sizes are small
+% then step vectors are increased
+done   = false;            % is some finishing criterion fulfilled?
+
+nVec   = 2*dim + 2;            % maximum index in step matrix
+opp_j  = @(j) f_opp_j(j,nVec); % short for opposite index in step matrix
+
+% init meta variables
+exitflag  = -1;        % flag indicating exit reason
+starttime = cputime;   % to measure time difference
+funEvals  = 0;         % function evaluations, should be <= maxFunEvals
+
+% init x, fval
+ybst      = y0;
+fbst      = fun(ybst,funEvals);
+funEvals  = funEvals + 1;
+
+if (visual)
+    f_output(denormalize(ybst),fbst,funEvals,'init',outputFcn); % create new figure and initialize
+    f_output(denormalize(ybst),fbst,funEvals,'iter',outputFcn); % first iteration with start point and jIter = 0
+end
+
+while (~done)
```

Comment:
100 line loop body, 5 levels of indentation. can't follow that :crying_cat_face:

Reply:
@yannikschaelte : Could you maybe rearrange this part of the code? I think Daniel is right here, it is kind of hard to read...
You could maybe use something like a struct for the current optimizer state, in which different variables can be grouped. Then it would be easier to transfer code parts to subfunctions, as you could pass maybe 3 instead of 10 arguments...

Reply:
improved readability a bit, added comments, reduced passed parameters. however, it is also good to see which values exactly are passed to subfunctions so as to know what the decision the subfunction makes will depend on.
concerning the length of the main function: i don't think it would help to move parts to subfunctions, it's simply a long and intertwined decision structure.
can try it though if you think it'd be better.